### PR TITLE
feat(BUX-397): acceptance and rejection of establishing contacts

### PR DESCRIPTION
**Pull request by jakubmkowalski**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: actions/contacts/reject.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	paymail := c.Param("paymail")
+
+	err := a.Services.SpvWalletEngine.RejectContact(c, reqXPubID, paymail)
+
+	if err != nil {
+		c.JSON(http.StatusExpectationFailed, err.Error())
+		return
+	}
+
+	c.JSON(http.StatusOK, "contact rejected")
```
Comment (arkadiuszos4chain) [inline actions/contacts/reject.go:32]:
```suggestion
	c.Status(http.StatusOK)
```

File: actions/contacts/reject.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// @Description	Reject contact. For contact with status "awaiting" delete contact
+// @Tags		Contact
+// @Produce		json
+// @Param		paymail query string true "paymail"
+// @Success		200
+// @Router		/v1/contact/reject [PATCH]
+// @Security	x-auth-xpub
+func (a *Action) reject(c *gin.Context) {
+	reqXPubID := c.GetString(auth.ParamXPubHashKey)
+	// paymail := c.Query("paymail")
```
Comment (arkadiuszos4chain) [inline actions/contacts/reject.go:22]: Remove comment

File: engine/action_contact_test.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
+package engine
```
Comment (arkadiuszos4chain) [inline engine/action_contact_test.go:1]: Please replace classic error checking with `require.NoError(t, err)`

File: actions/contacts/upsert.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ import (
 // @Produce		json
 // @Param		paymail path string true "Paymail address of the contact the user wants to add/modify"
 // @Param		UpsertContact body contacts.UpsertContact true "Full name and metadata needed to add/modify contact"
+// @Param		paymail path string true "paymail"
```
Comment (arkadiuszos4chain) [inline actions/contacts/upsert.go:23]: Please remove

Comment (jakubmkowalski) [reply actions/contacts/upsert.go]: removed

File: engine/action_contact_test.go (hunk 3 above)
Comment (jakubmkowalski) [reply engine/action_contact_test.go:1]: replaced

File: actions/contacts/reject.go (hunk 2 above)
Comment (jakubmkowalski) [reply actions/contacts/reject.go]: removed

File: actions/contacts/reject.go (hunk 1 above)
Comment (jakubmkowalski) [reply actions/contacts/reject.go]: changed

File: actions/contacts/accept.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// @Success		200
+// @Router		/v1/contact/accept [PATCH]
+// @Security	x-auth-xpub
+func (a *Action) accept(c *gin.Context) {
+	reqXPubID := c.GetString(auth.ParamXPubHashKey)
+	paymail := c.Param("paymail")
+
+	err := a.Services.SpvWalletEngine.AcceptContact(c, reqXPubID, paymail)
+
+	if err != nil {
```
Comment (arkadiuszos4chain) [inline actions/contacts/accept.go:26]: As we discussed, please return different HTTP status codes for different errors:
-    Contact not found: 404
-    Contact is in invalid status: 422
-    Other errors: 500

File: actions/contacts/reject.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+// reject will reject contact request
+// Reject contact godoc
+// @Summary		Reject contact
+// @Description	Reject contact. For contact with status "awaiting" delete contact
+// @Tags		Contact
+// @Produce		json
+// @Param		paymail query string true "paymail"
+// @Success		200
+// @Router		/v1/contact/reject [PATCH]
```
Comment (wregulski) [inline actions/contacts/reject.go:18]: Hmm, I think that something doesn't fit together:
`// @Param		paymail query string true "paymail" `

Here you introduced the query param, but in `routes.go` it looks that it is rather a path param:
`contactGroup.PATCH("/reject/:paymail", action.reject)`

The second thing is that the endpoint should be named as:
`/v1/contact/rejected`
and
`/v1/contact/accepted`

So base on our previous team discussion:
image

Comment (wregulski) [reply actions/contacts/reject.go:18]: ### To sum it up and be consistent with our arrangements:

1. Rename route paths to:
`/v1/contact/rejected/:paymail`
and
`/v1/contact/accepted/:paymail`

2. Stick with path params in swagger:
`// @Param		paymail ~~query~~ string true "paymail"`
`// @Param		paymail path string true "paymail"`

and add this param to the route in swagger:
`// @Router		/v1/contact/rejected/{paymail}`
`// @Router		/v1/contact/accepted/{paymail}`

Please regenerate swagger after those changes.

File: engine/action_contact.go [hunk 7]
```diff
@@ -213,3 +214,47 @@ func (c *Client) GetContacts(ctx context.Context, metadata *Metadata, conditions
...
+
+	contact, err := getContact(ctx, paymail, xPubID, c.DefaultModelOptions()...)
+	if err != nil {
+		return err
+	}
+	if contact == nil {
+		return errors.New("contact not found")
+	}
+	if contact.Status != ContactAwaitAccept {
+		return errors.New("contact does not have status awaiting")
```
Comment (wregulski) [inline engine/action_contact.go:228]: let's include current status in this error

**Review by wregulski — CHANGES_REQUESTED** [review]

File: actions/contacts/reject.go (hunk 6 above)
Comment (arkadiuszos4chain) [reply actions/contacts/reject.go:18]: I think it in the scope of https://bitcoinassociation.atlassian.net/jira/software/c/projects/BUX/boards/22?selectedIssue=BUX-648

Comment (wregulski) [reply actions/contacts/reject.go:18]: @arkadiuszos4chain I would vote to do this now, why to not have properly adjusted endpoints if it's a fresh PR?

File: engine/action_contact.go [hunk 8]
```diff
@@ -213,3 +216,55 @@ func (c *Client) GetContacts(ctx context.Context, metadata *Metadata, conditions
...
+
+func (c *Client) RejectContact(ctx context.Context, xPubID, paymail string) error {
+
+	contact, err := getContact(ctx, paymail, xPubID, c.DefaultModelOptions()...)
+	if err != nil {
+		c.Logger().Err(err).Msgf("unexpected error while geting contact for paymail %s and xPubID %s", paymail, xPubID)
+		return err
+	}
+	if contact == nil {
+		c.Logger().Err(ErrContactNotFound).Msgf("contact not found for paymail %s and xPubID %s", paymail, xPubID)
```
Comment (arkadiuszos4chain) [inline engine/action_contact.go:252]: We don't need this log

File: engine/action_contact.go [hunk 9]
```diff
@@ -213,3 +216,55 @@ func (c *Client) GetContacts(ctx context.Context, metadata *Metadata, conditions
...
+		return ErrContactNotFound
+	}
+	if contact.Status != ContactAwaitAccept {
+		err = ErrContactStatusNotAwaiting
```
Comment (arkadiuszos4chain) [inline engine/action_contact.go:256]: IMO it should be Warning

Comment (arkadiuszos4chain) [reply engine/action_contact.go:256]: Please add xpub and paymail to log context information

File: engine/action_contact.go [hunk 10]
```diff
@@ -213,3 +216,55 @@ func (c *Client) GetContacts(ctx context.Context, metadata *Metadata, conditions
...
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) RejectContact(ctx context.Context, xPubID, paymail string) error {
+
+	contact, err := getContact(ctx, paymail, xPubID, c.DefaultModelOptions()...)
+	if err != nil {
```
Comment (arkadiuszos4chain) [inline engine/action_contact.go:247]: Please move xpub and paymail to log context information

File: engine/action_contact.go [hunk 11]
```diff
@@ -213,3 +216,55 @@ func (c *Client) GetContacts(ctx context.Context, metadata *Metadata, conditions
...
+
+func (c *Client) AcceptContact(ctx context.Context, xPubID, paymail string) error {
+
+	contact, err := getContact(ctx, paymail, xPubID, c.DefaultModelOptions()...)
+	if err != nil {
+		c.Logger().Err(err).Msgf("unexpected error while geting contact for paymail %s and xPubID %s", paymail, xPubID)
+		return err
+	}
+	if contact == nil {
+		c.Logger().Err(ErrContactNotFound).Msgf("contact not found for paymail %s and xPubID %s", paymail, xPubID)
```
Comment (arkadiuszos4chain) [inline engine/action_contact.go:228]: We don't need this log

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

**Review by arkadiuszos4chain — APPROVED** [review]

**Review by wregulski — APPROVED** [review]

File: engine/action_contact.go (hunk 7 above)
Comment (jakubmkowalski) [reply engine/action_contact.go]: included

File: engine/action_contact.go (hunk 8 above)
Comment (jakubmkowalski) [reply engine/action_contact.go]: removed

File: engine/action_contact.go (hunk 9 above)
Comment (jakubmkowalski) [reply engine/action_contact.go:263]: changed to warning, xpub and paymail moved to context

File: engine/action_contact.go (hunk 11 above)
Comment (jakubmkowalski) [reply engine/action_contact.go]: removed

File: engine/action_contact.go (hunk 10 above)
Comment (jakubmkowalski) [reply engine/action_contact.go:252]: xpub and paymail moved to context

**Comment by arkadiuszos4chain** [comment]
testing

**Pull request merged**